docs/concepts.rst

```diff
@@ -318,49 +318,114 @@ argument of most object-specific I/O methods. You can also use
-            key = parsed.path[1:]
-            self.s3.Object(bucket, key).put(Body=txt, ContentEncoding="utf-8")
-         else:
-            super().write_text(dest, txt, *args, **kwargs)
-
-   StacIO.set_default(CustomStacIO)
-
+<https://boto3.amazonaws.com/v1/documentation/api/latest/index.html>`__
+or Azure Blob Storage using the `Azure SDK for Python
+<https://learn.microsoft.com/en-us/python/api/overview/azure/storage-blob-readme?view=azure-python>`__:
+
+.. tab-set::
+   .. tab-item:: AWS S3
+
+      .. code-block:: python
+
+         from urllib.parse import urlparse
+         import boto3
+         from pystac import Link
+         from pystac.stac_io import DefaultStacIO, StacIO
+         from typing import Union, Any
+
+         class CustomStacIO(DefaultStacIO):
+            def __init__(self):
+               self.s3 = boto3.resource("s3")
+               super().__init__()
+
+            def read_text(
+               self, source: Union[str, Link], *args: Any, **kwargs: Any
+            ) -> str:
+               parsed = urlparse(source)
+               if parsed.scheme == "s3":
+                  bucket = parsed.netloc
+                  key = parsed.path[1:]
+
+                  obj = self.s3.Object(bucket, key)
+                  return obj.get()["Body"].read().decode("utf-8")
+               else:
+                  return super().read_text(source, *args, **kwargs)
+
+            def write_text(
+               self, dest: Union[str, Link], txt: str, *args: Any, **kwargs: Any
+            ) -> None:
+               parsed = urlparse(dest)
+               if parsed.scheme == "s3":
+                  bucket = parsed.netloc
+                  key = parsed.path[1:]
+                  self.s3.Object(bucket, key).put(Body=txt, ContentEncoding="utf-8")
+               else:
+                  super().write_text(dest, txt, *args, **kwargs)
+
+         StacIO.set_default(CustomStacIO)
+
+   .. tab-item:: Azure Blob Storage
+
+      .. code-block:: python
+
+         import os
+         from pathlib import PurePosixPath
+         from typing import Any, Tuple, Union
+         from urllib.parse import urlparse
+
+         from azure.storage.blob import BlobClient, ContentSettings
+         from pystac import Link
+         from pystac.stac_io import DefaultStacIO, StacIO
+
+         class BlobStacIO(DefaultStacIO):
+            """A custom StacIO class for reading and writing STAC objects
+            from/to Azure Blob storage.
+            """
+
+            def _parse_blob_url(self, url: str) -> Tuple[str, str]:
+               path = PurePosixPath(urlparse(url).path)
+               container = path.parts[1]
+               blob = "/".join(path.parts[2:])
+               return container, blob
+
+            def _get_blob_client(self, container: str, blob: str) -> BlobClient:
+               return BlobClient.from_connection_string(
+                     os.environ["AZURE_STORAGE_CONNECTION_STRING"],
```

Comment:
I'm not a huge fan of implicit dependencies on env vars in the examples — could we define this connection string as a class-level attribute that falls back to an env var?

Reply:
Good call!

Reply:
There are a number of ways to instantiate the `BlobClient` depending on how you want to authenticate. I updated the example to support either the `conn_str` or `account_url` + `credential` if required. These are now class attributes and the example includes a section showing how to set them.